### PR TITLE
Support query attributes for vector queries

### PR DESCRIPTION
**Pull request by alonre24**
Allow using query attributes syntax for vector queries, for specifying both vector search related attributes and the vector distance field. That is, both syntax are allowed: `...=>[KNN 10 @v $b EF_RUNTIME 100 AS vec_dist]` and `...=>[KNN 10 @v $b]=>{$EF_RUNTIME:100; $YIELD_DISTANCE_AS:vec_dist}`. The `YIELD_DISTANCE_AS` is allowed for every query nodes whose (new) QueryNode_YieldsDistance flag is set to true (only vector node currently).
*** This includes a fix - allowing query attributes receive a stop-word as value (was a syntax error before)
(was split from #3043)

### CODE REVIEW

File: src/query_parser/v2/parser.y [hunk 1]
```diff
@@ -504,7 +504,22 @@ attribute(A) ::= ATTRIBUTE(B) COLON param_term(C). {
       value_len = found_value_len;
     }
   }
-  A = (QueryAttribute){ .name = B.s, .namelen = B.len, .value = value, .vallen = value_len };
+  A = (QueryAttribute){ .name = rm_strndup(B.s, B.len), .namelen = B.len, .value = value, .vallen = value_len };
+}
+
+attribute(A) ::= ATTRIBUTE(B) COLON STOPWORD(C). {
```
Comment (ashtul) [inline src/query_parser/v2/parser.y:510]: Let's avoid the stopword necessity by using strcmp instead of the compiler. like in attribute.

File: src/query_parser/v2/parser.y [hunk 2]
```diff
@@ -944,7 +959,59 @@ vector_score_field(A) ::= as STOPWORD(B). {
   A.type = QT_TERM;
 }
 
-// Every vector query will have basic command part. Right now we only have KNN command.
+// Use query attributes syntax
+query ::= expr(A) ARROW LSQB vector_query(B) RSQB ARROW LB attribute_list(C) RB. {
```
Comment (ashtul) [inline src/query_parser/v2/parser.y:963]: How is this related to the feature?

File: tests/cpptests/test_cpp_query.cpp [hunk 3]
```diff
@@ -475,10 +479,16 @@ TEST_F(QueryTest, testParser_v2) {
   assertValidQuery("*=>[KNN $K @vec_field $BLOB AS score]", ctx);
   assertValidQuery("*=>[KNN $K @vec_field $BLOB EF $ef foo bar x 5 AS score]", ctx);
   assertValidQuery("*=>[KNN $K @vec_field $BLOB foo bar x 5]", ctx);
+  // Using query attributes syntax is also allowed.
+  assertValidQuery("*=>[knn $K @vec_field $BLOB]=>{$yield_distance_as: vec_dist;}", ctx);
```
Comment (ashtul) [inline tests/cpptests/test_cpp_query.cpp:483]: Please add tests for all field type with errors

File: src/query.c [hunk 4]
```diff
@@ -292,6 +296,7 @@ QueryNode *NewVectorNode_WithParams(struct QueryParseCtx *q, VectorQueryType typ
   VectorQuery *vq = rm_calloc(1, sizeof(*vq));
   ret->vn.vq = vq;
   vq->type = type;
+  ret->opts.flags |= QueryNode_YieldsDistance;
```
Comment (ashtul) [inline src/query.c:299]: Why do we need this flag?

File: src/query.c [hunk 5]
```diff
@@ -830,6 +835,24 @@ static IndexIterator *Query_EvalVectorNode(QueryEvalCtx *q, QueryNode *qn) {
   if (!fs || !FIELD_IS(fs, INDEXFLD_T_VECTOR)) {
     return NULL;
   }
+
+  if (qn->opts.distField) {
+    // Since the KNN syntax allows specifying the distance field in two ways (...=>[KNN ... AS <dist_field>] and
+    // ...=>[KNN ...]=>{$YIELD_DISTANCE_AS:<dist_field>), we validate that we got it only once.
+    char default_score_field[strlen(qn->vn.vq->property) + 9]; // buffer for __<field>_score
+    sprintf(default_score_field, "__%s_score", qn->vn.vq->property);
```
Comment (ashtul) [inline src/query.c:843]: I don't recall using sprintf in RediSearch. Why do we need it here?

File: src/query.c [hunk 6]
```diff
@@ -1753,63 +1776,91 @@ int QueryNode_ForEach(QueryNode *q, QueryNode_ForEachCallback callback, void *ct
...
 
-  } else {
-    QueryError_SetErrorFmt(status, QUERY_ENOOPTION, "Invalid attribute %.*s", (int)attr->namelen,
-                           attr->name);
-    return 0;
+  } else if (STR_EQCASE(attr->name, attr->namelen, YIELD_DISTANCE_ATTR) && qn->opts.flags & QueryNode_YieldsDistance) {
+    qn->opts.distField = rm_strndup(attr->value, attr->vallen);
+    res = 1;
+
+  } else if (qn->type == QN_VECTOR) {
```
Comment (ashtul) [inline src/query.c:1856]: Why do we need to limit this for vector only. 
We can set the flag regardless and just ignore it if not supported.

File: coord/src/module.c [hunk 7]
```diff
@@ -492,7 +492,7 @@ void prepareOptionalTopKCase(searchRequestCtx *req, RedisModuleString **argv, in
   if(queryNode!= NULL && queryNode->type == QN_VECTOR) {
     QueryVectorNode queryVectorNode = queryNode->vn;
     size_t k = queryVectorNode.vq->knn.k;
-    const char* scoreField = queryVectorNode.vq->scoreField;
+    const char* scoreField = queryNode->opts.distField ? queryNode->opts.distField : queryVectorNode.vq->scoreField;
```
Comment (ashtul) [inline coord/src/module.c:495]: Why is this needed?

**Review by ashtul — CHANGES_REQUESTED** [review]

File: src/query_parser/v2/parser.y (hunk 1 above)
Comment (alonre24) [reply src/query_parser/v2/parser.y:510]: I'm not sure how to do that, let's discuss it...

File: src/query.c (hunk 4 above)
Comment (alonre24) [reply src/query.c:299]: This is part of the generic yield distance mechanism, the purpose is that every query node which is able to yield a distance field, will set this flag to true. We can discuss this approach further if you want.

File: src/query.c (hunk 5 above)
Comment (alonre24) [reply src/query.c:843]: Its for supporting an edge case - with the new query attributes syntax, one can define the distance field twice as following:
`*=>[KNN 10 @v $b AS dist1]=>{$yield_distance_as:dist2}`. Hence, we return an error if the distance field was defined twice. However, in cases where distance field is defined only using the query attribute, we set `qn->vn.vq->scoreField` to be the default `__v_score`. In that case only, we don't return an error, but override the default score field.
`sprintf` is used to avoid allocations (save the default score field on the stack instead)

File: src/query.c (hunk 6 above)
Comment (alonre24) [reply src/query.c:1856]: Note that in `QueryVectorNode_ApplyAttribute` we set the vector attributes into `VectorQueryParams` struct, which is part of a vector node only. To follow your suggestion, I guess we should add the vector attributes to the generic `QueryNodeOptions`. Let's discuss it...

File: coord/src/module.c (hunk 7 above)
Comment (alonre24) [reply coord/src/module.c:495]: If we have `queryNode->opts.distField` set, then user gave a distance field as a query attribute. Otherwise, either the user specified a distance field name in the `AS` clause (`[KNN ... AS v_dist]`), or we have the default field stored in `queryVectorNode.vq->scoreField`. As I explained [here](https://github.com/RediSearch/RediSearch/pull/3105#discussion_r977641769), if both fields are not NULL, we get an error anyway.

File: src/query_parser/v2/parser.y (hunk 1 above)
Comment (ashtul) [reply src/query_parser/v2/parser.y:510]: We can use the current mechanism using `strcmp`.
If the decision is to move away from it, then your code stands.

File: src/query.c (hunk 5 above)
Comment (ashtul) [reply src/query.c:843]: Let's have @filipecosta90 test the performance impact.

File: coord/src/module.c (hunk 7 above)
Comment (ashtul) [reply coord/src/module.c:495]: We can have a `distField` in addition to the `scoreField`.
In the future, if we want to calculate something both, we will have it already.

File: src/query.c (hunk 5 above)
Comment (ashtul) [reply src/query.c:843]: I see we use it all over the project.

**Review by ashtul — DISMISSED** [review]
Approved after discussion

File: src/query_parser/v2/parser.y (hunk 1 above)
Comment (alonre24) [reply src/query_parser/v2/parser.y:510]: solved by #3130

File: tests/cpptests/test_cpp_query.cpp [hunk 8]
```diff
@@ -491,15 +501,19 @@ TEST_F(QueryTest, testParser_v2) {
   assertInvalidQuery("*=>[KNN $K @vec_field $BLOB $EF ef foo bar x 5 AS score]", ctx); // parameter as attribute
   assertInvalidQuery("*=>[KNN $K @vec_field $BLOB EF ef foo bar x 5 AS ]", ctx); // not specifying score field name
   assertInvalidQuery("*=>[KNN $K @vec_field $BLOB EF ef foo bar x]", ctx); // missing parameter value (passing only key)
+  assertInvalidQuery("*=>[KNN $K @vec_field $BLOB => {$yield:dist}]", ctx); // invalid attributes syntax
+  assertInvalidQuery("*=>[KNN $K @vec_field $BLOB EF_RUNTIME 100 => {$yield_distance_as:dist;}]", ctx); // invalid combined syntax
```
Comment (DvirDukhan) [inline tests/cpptests/test_cpp_query.cpp:505]: Why is it either? Why we can't mix?

File: src/query.c [hunk 9]
```diff
@@ -1753,63 +1776,91 @@ int QueryNode_ForEach(QueryNode *q, QueryNode_ForEachCallback callback, void *ct
   return retVal;
 }
 
+// Convert the query attribute into a raw vector param to be resolved by the vector iterator
+// down the road. return 0 in case of an unrecognized parameter.
+static int QueryVectorNode_ApplyAttribute(VectorQuery *vq, QueryAttribute *attr) {
+  if (STR_EQCASE(attr->name, attr->namelen, VECSIM_EFRUNTIME) ||
+      STR_EQCASE(attr->name, attr->namelen, VECSIM_HYBRID_POLICY) ||
+      STR_EQCASE(attr->name, attr->namelen, VECSIM_BATCH_SIZE)) {
+    VecSimRawParam param = (VecSimRawParam){ .name = rm_strndup(attr->name, attr->namelen),
```
Comment (DvirDukhan) [inline src/query.c:1785]: Why do we need strdup here? The raw param is `const char *`

File: src/query.c [hunk 10]
```diff
@@ -1753,63 +1776,91 @@ int QueryNode_ForEach(QueryNode *q, QueryNode_ForEachCallback callback, void *ct
...
+
+  } else if (qn->type == QN_VECTOR) {
+    res = QueryVectorNode_ApplyAttribute(qn->vn.vq, attr);
   }
 
-  return 1;
+  if (res == 1) return res;
+  QueryError_SetErrorFmt(status, QUERY_ENOOPTION, "Invalid attribute %.*s", (int)attr->namelen,
+                         attr->name);
+  return 0;
```
Comment (DvirDukhan) [inline src/query.c:1863]:
```suggestion
  if (!res) {
  QueryError_SetErrorFmt(status, QUERY_ENOOPTION, "Invalid attribute %.*s", (int)attr->namelen,
                         attr->name);
}
  return res;
```

File: src/query_parser/v2/parser.y [hunk 11]
```diff
@@ -502,7 +502,7 @@ attribute(A) ::= ATTRIBUTE(B) COLON param_term(C). {
       value_len = found_value_len;
     }
   }
-  A = (QueryAttribute){ .name = B.s, .namelen = B.len, .value = value, .vallen = value_len };
+  A = (QueryAttribute){ .name = rm_strndup(B.s, B.len), .namelen = B.len, .value = value, .vallen = value_len };
```
Comment (DvirDukhan) [inline src/query_parser/v2/parser.y:505]: why dup?

File: src/query_parser/v2/parser.y [hunk 12]
```diff
@@ -910,11 +910,64 @@ vector_query(A) ::= vector_command(B). {
...
+  }
+  ctx->root = B;
+    if (B && C) {
+       QueryNode_ApplyAttributes(B, C, array_len(C), ctx->status);
+    }
+    array_free_ex(C, rm_free((char*)((QueryAttribute*)ptr )->name); rm_free((char*)((QueryAttribute*)ptr )->value));
+
+  if (A) {
+    QueryNode_AddChild(B, A);
+  }
```
Comment (DvirDukhan) [inline src/query_parser/v2/parser.y:952]: nit: please be coherent with the similar logic above on where you are attaching the children. Also please add comment specifying the use cases (expr, text expr)

File: tests/pytests/test_vecsim.py [hunk 13]
```diff
@@ -376,11 +380,18 @@ def testSearchErrors():
...
 
+    # Invalid query combination with query attributes syntax.
+    env.expect('FT.SEARCH', 'idx', '*=>[KNN 2 @v $b AS score]=>{$yield_distance_as:score2;}', 'PARAMS', '2', 'b', 'abcdefgh').error().contains('Distance field was specified twice for vector query: score and score2')
+    env.expect('FT.SEARCH', 'idx', '*=>[KNN 2 @v $b EF_RUNTIME 100]=>{$EF_RUNTIME:200;}', 'PARAMS', '2', 'b', 'abcdefgh').error().contains('Error parsing vector similarity parameters: Parameter was specified twice')
+    env.expect('FT.SEARCH', 'idx', '*=>[KNN 2 @v $b AS $score_1]=>{$yield_distance_as:$score_2;}', 'PARAMS', '6', 'b', 'abcdefgh', 'score_1', 'score_1_val', 'score_2', 'score_2_val').error().contains('Distance field was specified twice for vector query: score_1_val and score_2_val')
+    env.expect('FT.SEARCH', 'idx', 'hello=>[KNN 2 @v $b AS score]=>{$yield_distance_as:__v_score;}', 'PARAMS', '2', 'b', 'abcdefgh').error().contains('Distance field was specified twice for vector query: score and __v_score')
```
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/pytests/test_vecsim.py:393]: Currently it's an error, except for the case of specifying the default score field name in both ways

File: tests/cpptests/test_cpp_query.cpp (hunk 8 above)
Comment (alonre24) [reply tests/cpptests/test_cpp_query.cpp:505]: We can mix, the syntax error in this test is that the `]` of the KNN clause should have come before the `=>`

File: src/query.c (hunk 9 above)
Comment (alonre24) [reply src/query.c:1785]: The caller function `QueryNode_ApplyAttributes` is called from the parser, where `QueryAttribute` struct is being constructed and **destructed**. Hence, we need to move the ownership for these string here. But you're right that we can just set `attr->name` and `attr->value` to be NULL instead of using `strndup`.

File: src/query_parser/v2/parser.y (hunk 11 above)
Comment (alonre24) [reply src/query_parser/v2/parser.y:505]: For vector attributes, we need the hold the name string in addition to the value string which is already held. But in second thought, we don't need the call `dup` here, but we can do it in `QueryVectorNode_ApplyAttribute` instead (and save `dup` for non-vector attributes). It will also allow us to revert the `rm_free((char*)$$.name);` calls that were added to free the inner name field in the `QueryAttribute` array...

File: src/query_parser/v2/parser.y (hunk 12 above)
Comment (alonre24) [reply src/query_parser/v2/parser.y:952]: Not sure I understand what you meant here...

**Review by DvirDukhan — APPROVED** [review]
